### ISSUE TITLE
Regarding PersistentStateAttribute support in custom storage provider implementation

### ISSUE THREAD

**Issue by nkosi23** [dotnet/orleans#7671]
I have implemented a custom storage provider using [the DynamoDB](https://github.com/dotnet/orleans/blob/main/src/AWS/Orleans.Persistence.DynamoDB/Provider/DynamoDBGrainStorage.cs) implementation as a blueprint. However I do not understand how the storage provider is supposed to provide support for multiple named state. For example [the documentation](https://docs.microsoft.com/en-us/dotnet/orleans/grains/grain-persistence/) provides the following sample:

```csharp
public class UserGrain : Grain, IUserGrain
{
    private readonly IPersistentState<ProfileState> _profile;
    private readonly IPersistentState<CartState> _cart;

    public UserGrain(
        [PersistentState("profile", "profileStore")] IPersistentState<ProfileState> profile,
        [PersistentState("cart", "cartStore")] IPersistentState<CartState> cart)
    {
        _profile = profile;
        _cart = cart;
    }
}
```

I am wondering if the following must be supported by storage providers:

```csharp
public class UserGrain : Grain, IUserGrain
{
    private readonly IPersistentState<ProfileState> _profile;
    private readonly IPersistentState<CartState> _cart;

    public UserGrain(
        [PersistentState("profile", "userStore")] IPersistentState<ProfileState> profile,
        [PersistentState("cart", "userStore")] IPersistentState<CartState> cart)
    {
        _profile = profile;
        _cart = cart;
    }
}
```

In this last sample, the same storage provider is used to load multiple states distinguished by their name only. Looking at the DynamoDB and Azure implementations, I am under the impression that storage providers do not support this use case since states are only identified in the underlying storage using the grain identity. Moreover no "stateName" parameter is passed to ReadStateAsync and WriteStateAsync. 

This surprised me as I assumed that this use case was supported so I just wanted to confirm my understanding: if I understand correctly, if one uses the built-in persistence facility, each state that a grain may have is expected to be stored in a different database (or bucket, or table...). Therefore where and how is the stateName parameter of the PersistentStateAttribute used?

**Comment by JorgeCandeias** [comment]
TLDR; Providers can ignore shard logic, all they need to care about is a name and an object.

Shard support is implemented at a higher level, in the [PersistentStateFactory](https://github.com/dotnet/orleans/blob/afa8d6295b59983ab8985774f1578e773b0a7992/src/Orleans.Runtime/Facet/Persistent/PersistentStateStorageFactory.cs#L35).

Click through the link and you'll see a line that just concatenates the grain id with the shard name defined by the `PersistentStateAttribute` using a dot, which implements `IPersistentStateConfiguration`. The factory then [gives this full name](https://github.com/dotnet/orleans/blob/afa8d6295b59983ab8985774f1578e773b0a7992/src/Orleans.Runtime/Facet/Persistent/PersistentStateStorageFactory.cs#L28) to the internal `PersistentStateBridge`, which then passes it on to the underlying storage implementation.

The factory is mapped to the attribute via the [PersistentStateAttributeMapper](https://github.com/dotnet/orleans/blob/afa8d6295b59983ab8985774f1578e773b0a7992/src/Orleans.Runtime/Facet/Persistent/PersistentStateAttributeMapper.cs).

All together they make up the facet pattern, which is an advanced pattern that Orleans supports, and works the same way as parameter bindings in Azure Functions. You can roll your own facets as well, by the way.

So the specific storage provider implementation does not have to do anything in particular to support shards, as the name given to it for a specific state item already includes the shard name.

**Comment by nkosi23** [comment]
Thank you so much for these insights! This clarifies a lot of things in my mind, especially when it comes to the facet pattern as I do not think I came across it before. It is not listed in the GoF patterns apparently, advanced material as you say, Microsoft is Microsoft for a reason :smile:. I will definitely study this pattern more deeply, it seems to be used a lot in Orleans and it could be useful even in my applications ([found this](https://hillside.net/plop/plop2002/final/plop2002_ecrahen0_0.pdf)). Many thanks!

I'd just have one last question: in the code you highlighted I can see that `PersistentStateBridge<TState>` [passes](https://github.com/dotnet/orleans/blob/afa8d6295b59983ab8985774f1578e773b0a7992/src/Orleans.Runtime/Facet/Persistent/PersistentStateStorageFactory.cs#L101) the `fullStateName` to the constructor of `StateStorageBridge<TState>`. Going down the rabbit hole, I can see that this name [is then passed](https://github.com/dotnet/orleans/blob/f9299762e433ed2b426f7a6e676776d224f13f52/src/Orleans.Runtime/Storage/StateStorageBridge.cs#L70) as the first parameter of `IGrainStorage.WriteStateAsync`, `IGrainStorage.ReadStateAsync` and `IGrainStorage.ClearStateAsync`. What confuses me is that [the definition of IGrainStorage](https://github.com/dotnet/orleans/blob/f9299762e433ed2b426f7a6e676776d224f13f52/src/Orleans.Core/Providers/IGrainStorage.cs#L20),  says that the first parameter is called `string grainType` and its description reads:

> Type of this grain [fully qualified class name]

 Therefore, based on your explanation and demonstration, I am tempted to conclude that it is safe to consider that this XML documentation is wrong and that I can safely consider that the first parameter of the `IGrainStorage.X` methods actually is the name of the state and not the fully qualified type of the grain. Is that correct?

**Comment by JorgeCandeias** [comment]
> Therefore, based on your explanation and demonstration, I am tempted to conclude that it is safe to consider that this XML documentation is wrong and that I can safely consider that the first parameter of the `IGrainStorage.X` methods actually is the name of the state and not the fully qualified type of the grain. Is that correct?

Yes, it looks outdated. Like with any other cutting edge technology, the code is always further ahead of the docs by a few years.

 `Grain<T>` and `IGrainStorage` came before, `IPersistentState<T>` only came some time after that, and it was built on top of the existing code, in a way that would not break existing persistence providers. I suspect that's why `PersistentStateBridge<T>` exists, as it's a compatibility class whose only work is to let the persistence facet play nice with the core system and published plugins.

Note that persistence providers are free to alter the state name further. For example the ADO.NET provider [further formats the state name](https://github.com/dotnet/orleans/blob/a2b808721193f313f1d67c1b2ea815783591ebfd/src/AdoNet/Orleans.Persistence.AdoNet/Storage/Provider/AdoNetGrainStorage.cs#L344) into something that ends up looking like `{GrainType},{GrainAssembly}.{StateName}` on the `GrainTypeString` column. I don't know *why* the ado provider does this though, it looks unnecessary, but there's likely an historical reason for that as well.
